[iOS] Text selection is lost if right clicking at end of selection

rr-codes opened the pull request:
#### 347a3c86979f5d5873d85587fa1d7e78b75106ec
<pre>
[iOS] Text selection is lost if right clicking at end of selection
<a href="https://bugs.webkit.org/show_bug.cgi?id=256126">https://bugs.webkit.org/show_bug.cgi?id=256126</a>
rdar://105737822

Reviewed by Wenson Hsieh.

In WebKit, a point is considered to be inside a selection if the boundary point created from the
point is within the boundary points of the selection range, inclusive. For example, with the text
&quot;Hello&quot; and a selection of &quot;Hell&quot;, if the point is slightly to the left of the middle of the &quot;o&quot;,
the boundary point of the point is considered to be the same as the boundary point of the end of
the selection.

On macOS, when right clicking at the end of a selection, if the clicked point is considered to be
&quot;inside&quot; the selection, a new selection is not created. However, for iOS, there is no such mechanism
to check this, so a new selection is always created, which is unexpected behavior if the click is
still &quot;inside&quot; the existing selection.

This PR fixes this by implementing this check on iOS.

* Source/WebCore/editing/FrameSelection.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::prepareSelectionForContextMenuWithLocationInView):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/iOSMouseSupport.mm:
(simulateEditContextMenuAppearance):
(TEST):

Canonical link: <a href="https://commits.webkit.org/263536@main">https://commits.webkit.org/263536@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9d17627f290ca8f1edfa9c080069bc02b8ce3b6a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4932 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5058 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5230 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6449 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/5038 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/4926 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/5233 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5030 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/5281 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5018 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5094 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4419 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/6465 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/2582 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/4417 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/9375 "1 flakes 145 failures") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/4445 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4489 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/6082 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4901 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3999 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4400 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1199 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8468 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4762 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -2477,13 +2477,20 @@ static inline bool rectIsTooBigForSelection(const IntRect& blockRect, const Loca
     if (!hitNode)
         return completionHandler(false, { });
 
+    if (auto* view = frame->view()) {
```
```suggestion
    if (RefPtr view = frame->view()) {
```

rr-codes replied:
@whsieh why is this better?

whsieh replied:
Generally speaking, we should use `Ref` and `RefPtr` when storing references to ref-counted objects on the stack.

More specifically, when using `auto* view = …`, `view` is just a raw pointer on the stack; this might not be an issue if we're only using trivial getters to read state on the frame view, but since `FrameSelection::contains` triggers layout (via hit-testing), there's no guarantee that the `view` pointer is valid through its entire lifetime. You could imagine that if we added code later on that referenced `view` after calling `frame->selection().contains(pointInContents)`, we would have a security bug (UAF).

whsieh commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -2477,13 +2477,20 @@ static inline bool rectIsTooBigForSelection(const IntRect& blockRect, const Loca
     auto* localMainFrame = dynamicDowncast<LocalFrame>(m_page->mainFrame());
     if (!localMainFrame)
-        return;
+        return completionHandler(false, { });
```
Good catch!

whsieh commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -2477,13 +2477,20 @@ static inline bool rectIsTooBigForSelection(const IntRect& blockRect, const Loca
         return completionHandler(false, { });
 
+    if (auto* view = frame->view()) {
+        LayoutPoint vPoint = view->windowToContents(point);
```
Hmm... is `windowToContents` the right transformation here? The `point` comes directly from the UI process via UIKit code that works with `WKContentView`, so I believe it's in root view coordinates, in which case `view->rootViewToContents(point)` should be preferable.

That said, it's possible that since `frame` is only ever the main frame here, it doesn't actually make a difference.

Nit - instead of `vPoint`, we should write out [full words unless it's a well-known abbreviation](https://webkit.org/code-style-guidelines/#names-full-words), and rename it to something like `viewPoint` or (perhaps more accurate) `pointInContents`.

rr-codes replied:
Can you tell I just copy-pasted this line 😛 Will fix, thanks!

rr-codes committed:
[iOS] Text selection is lost if right clicking at end of selection
https://bugs.webkit.org/show_bug.cgi?id=256126
rdar://105737822

Reviewed by Wenson Hsieh.

In WebKit, a point is considered to be inside a selection if the boundary point created from the
point is within the boundary points of the selection range, inclusive. For example, with the text
"Hello" and a selection of "Hell", if the point is slightly to the left of the middle of the "o",
the boundary point of the point is considered to be the same as the boundary point of the end of
the selection.

On macOS, when right clicking at the end of a selection, if the clicked point is considered to be
"inside" the selection, a new selection is not created. However, for iOS, there is no such mechanism
to check this, so a new selection is always created, which is unexpected behavior if the click is
still "inside" the existing selection.

This PR fixes this by implementing this check on iOS.

* Source/WebCore/editing/FrameSelection.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::prepareSelectionForContextMenuWithLocationInView):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/iOSMouseSupport.mm:
(simulateEditContextMenuAppearance):
(TEST):

Canonical link: https://commits.webkit.org/263536@main
